Popover utils and auto-show

b-kelly opened the pull request:
This is a fork of #462 which has some incredibly useful content, but has been sitting too long over an implementation regression.

Added features:

* `Stacks.showPopover`, `Stacks.hidePopover`, `Stacks.attachPopover`, `Stacks.detachPopover` helper methods
* `data-s-popover-auto-show` attribute
* updated Popover documentation page

95% of the code from this PR has been straight copied from the existing PR. The only changes are as follows:

* auto-show now doesn't hook up document events like .is-visible
  * This is less than ideal (especially for new functionality), but I'll have a quick follow-up PR to add the `hide-on-outside-click` functionality from the previous PR
* minor readability tweaks to `getPopover` internal function
* Didn't carry over `hide-on-outside-click` or `TokenList`
  * both of these are coming in a follow-up PR (I wanted to get this smaller set of changes out first)

b-kelly committed:
add show/hidePopper helpers
add attach/detachPopper helpers
add auto-show attribute to use in place if is-visible

b-kelly committed:
add docs for new helpers/attribute

b-kelly committed:
update built files

bnickel commented on lib/ts/controllers/s-popover.ts:
```diff
@@ -315,6 +324,177 @@ namespace Stacks {
+        const { isPopover, controller, referenceElement, popover } = getPopover(element);
+
+        // Hide the popover so its events fire.
+        if (controller) { controller.hide(); }
```
```suggestion
        controller?.hide();
```

Maybe shorthand here?

bnickel commented on lib/ts/controllers/s-popover.ts:
```diff
@@ -315,6 +324,177 @@ namespace Stacks {
+        if (controller) { controller.hide(); }
+
+        // Remove the popover if it exists
+        if (popover) { popover.remove(); }
```
```suggestion
        popover?.remove();
```

bnickel commented on lib/ts/controllers/s-popover.ts:
```diff
@@ -315,6 +324,177 @@ namespace Stacks {
+
+        // Remove the popover controller and the aria-controls attributes.
+        if (isPopover) {
+            element.removeAttribute("data-controller");
```
I know we're planning to handle this separately, but there's a lot of risk associated with this line.  (E.g. killing `data-controller="s-popover s-tooltip"`.)  We could do a slapdash `element.setAttribute("data-controller", element.getAttribute("data-controller")?.replace(/\bs-popover\b/, ''))` or warn about it in the documentation.

b-kelly replied:
I took the cheap way out with set/removeAttribute, but you're totally right. I implemented your toggleController code as a stopgap for a more full featured solution.

bnickel commented on lib/ts/controllers/s-popover.ts:
```diff
@@ -315,6 +324,177 @@ namespace Stacks {
+            referenceElement.insertAdjacentElement("afterend", popover);
+        }
+
+        element.setAttribute("data-controller", "s-popover");
```
Same

bnickel commented on lib/ts/controllers/s-popover.ts:
```diff
@@ -315,6 +324,177 @@ namespace Stacks {
+        }
+
+        if (!popoverId) {
+            popoverId = "--stacks-s-popover-" + Math.random().toString(36).substring(2, 10);
```
We have a similar function in `TooltipController.generateId()`.  Maybe we could add a base method on StacksController:

```
export function generateId() {
    return `--stacks-${this.identifier}-{Math.random().toString(36).substring(2, 10)}`;
}
```

b-kelly replied:
I'm punting on this one for now. I'd have to make the method static (since it is accessed in a helper here), in which case it won't have access to `this.identifier` and I'd have to hardcode the controller name somewhere for both types

bnickel reviewed (approved):
This looks good _but_ the `removeAttribute("data-controller")` stuff will continue to be a deal-breaker for cases where we're doing a temporary popover on top of an element that otherwise has a tooltip.  (These are the main cases where we would be using `detachPopover` today.)

I know you're looking at token list separately, but could we have a stopgap here? E.g., 

```typescript
function toggleController(el: Element, name: string, include: boolean) {
  var controllers = new Set(el.getAttribute('data-controller')?.split(/\s+/));
  if (include) { controllers.add(name); } else { controllers.delete(name); }
  el.setAttribute('data-controller', [...controllers.values()].join(' '))
}
```

bnickel commented on docs/product/components/popovers.html:
```diff
@@ -281,6 +287,97 @@
+        </div>
+    </div>
+    <script>
+        window.addEventListener("load", () => {
```
Ugh... This is only firing if you land on the page directly, not navigating to it.  I forget what the right approach is here.

aaronshekey replied:
Ah, right right right. Yep. Toss this into a JS file and load it with `defer`. It's really great as a consumer of the docs to load everything via AJAX but pretty much universally trolls every new feature demo we build, lol.

b-kelly replied:
Added to `feature.popovers.js`

bnickel commented on docs/product/components/popovers.html:
```diff
@@ -281,6 +287,97 @@
+    </div>
+</div>
+<script>
+    window.addEventListener("load", () => {
```
Same.  Should these move into a .js file?

b-kelly replied:
Added to `feature.popovers.js`

b-kelly committed:
Update lib/ts/controllers/s-popover.ts

Co-authored-by: Brian Nickel <bnickel@stackexchange.com>

b-kelly committed:
Update lib/ts/controllers/s-popover.ts

Co-authored-by: Brian Nickel <bnickel@stackexchange.com>

b-kelly committed:
tweak attach/detachPopover to use a new toggleController helper so we're not overwriting existing controller values

b-kelly commented:
@bnickel I addressed your concerns, especially the ones concerning the `data-controller` attributes. I already have a green check from you, but would you like to give it another lookover?

aaronshekey committed:
JS build

aaronshekey committed:
Move the JS examples to feature.popovers.js

This satisfies our weird JS-based navigation

aaronshekey committed:
Kill the close button

We only want to show explicit close buttons on auto-shown popovers

aaronshekey committed:
Use a menu in the example

aaronshekey commented on lib/ts/controllers/s-popover.ts:
```diff
@@ -315,6 +324,193 @@ namespace Stacks {
+        /**
+         * When true, the popover will appear immediately when the controller connects.
+         */
+        autoShow?: boolean;
```
How difficult would it be to enable:

``` js
/**
* When set, `data-s-popover-toggle-class` will be set to this value on the controller element.
*/
toggleClass?: string;
```

I gave it a whirl. It didn't work on the first click, but on subsequent clicks, it allowed adding `is-selected` as expected on the originating element.

b-kelly replied:
This one is tricky-ish. I left the original behavior intact for when the visible class was on a popover at initialization. In this case, the popover did _not_ have the toggle class added. That _should_ be fixed in #616, where we purposefully introduce the regression along with the new `hide-on-outside-click` interaction settings. I'm punting here purposefully and will double check it on #616.

aaronshekey committed:
Merge branch 'develop' into bkelly/popover-utils

aaronshekey committed:
Additional build to resolve conflicts

bnickel committed:
`load` was never firing

b-kelly committed:
commit compiled code
